File: blind 75 leetcode/Intervals/non_overlapping_intervals.py
"""
https://leetcode.com/problems/non-overlapping-intervals/submissions/
https://www.youtube.com/watch?v=nONCGxWoUfM&ab_channel=NeetCode`

Solved on 7/18/2023

Key Steps:
1) sort the intervals by start and end times
2) Iterate through the list of sorted intervals, while keeping track of the last
non overlapping interval. 

In order for two intervals to be non-overlapping, the previous end must always
be less than or equal to the current interval's start. 
Therefore, we check prevEnd < start to check for an overlap.

Whenever we find an overlapping interval, we increment
the count to show it's been removed. However if two intervals don't overlap, 
update the non overlapping interval since the last interval has ended, we want
to check for future overlaps using this new interval now.

3) when faced with two intervals that are overlapping, we choose to remove
the interval that ends later in order to potentially avoid an overlap with a future interval.

Meaning prevEnd > end

If the previous interval's end is greater than the current intervals end,
we remove the previous interval, incrementing the count to 1, and then the last non overlapping
interval becomes the current.

O(NLogN) time because of the sort, O(N) space to hold the sorted intervals

"""
class Solution:
    def eraseOverlapIntervals(self, intervals: List[List[int]]) -> int:   
        sortedIntervals = sorted(intervals, key = lambda x: [x[0], x[1]])
        lastNonOverlapping = sortedIntervals[0]
        count = 0
        for i in range(1, len(sortedIntervals)):
            start, end = sortedIntervals[i]
            prevStart, prevEnd = lastNonOverlapping
            # if the previous end is greater than the current start,
            # this is overlapping
            if prevEnd > start:
                # we want to remove the interval that ends later in order to potentially
                # avoid an overlap with a future interval, this will help us keep
                # the number of removals to a minimum
                if prevEnd > end:
                    lastNonOverlapping = sortedIntervals[i]   
                count += 1
            else:
                lastNonOverlapping = sortedIntervals[i]
        return count
                
                

File: node_js_ds_and_algos_implementations/intervals/non_overlapping_intervals.js
class Solution {
    /**
     * @param {number[][]} intervals
     * @return {number}
     */
    eraseOverlapIntervals(intervals) {
        /*
        Alternate solution 6/30/2026
        Greedy approach simplifies things here
        Instead of sorting by start time (like in most interval problems),
        sort by end time

        The reasoning is because any interval that comes earlier in the list
        is less likely to collide with future intervals, since it's an earlier
        end time. So by default, we would be removing less intervals this way (resulting 
        in the minimum)

        The overlapping check would instead be if prev.end > cur.start since it's sorted
        by end time

        Loop through intervals starting from 1 ... n
            If prevEnd > curStart
                update remove count, but keep prev the same, since
                it's a smaller end time, so we decrease the chances of conflicts
            else
                update prev to be the current interval

        Time: O(NLogN)
        Space: O(1)
        */

        // sort by end time
        intervals.sort((a,b) => {
            if (a[1] < b[1]){
                return -1
            }
            else if (a[1] > b[1]){
                return 1
            }
            return 0
        })
        let res = 0
        let prev = intervals[0]
        for (let i = 1; i < intervals.length; ++i){
            let [prevStart, prevEnd] = prev
            let [curStart, curEnd] = intervals[i]
            // overlapping
            if (prevEnd > curStart){
                res++
            }
            else {
                prev = intervals[i]
            }
        }
        return res
    }
}

/**
 * @param {number[][]} intervals
 * @return {number}
 */
var eraseOverlapIntervals = function(intervals) {
    /*
    1) sort the intervals by start time
    2) loop through intervals, keeping track of the last non overlapping interval
    3) Whenever the previous interval's end > current interval's start, this is overlapping. 
        - Trick: greedy approach
            in order to determine which interval to keep and which one to remove, we remove
            the interval that has the greater end time and keep the one that has the smaller end time. 
            This is to prevent this interval from
            potentially overlapping with future intervals.
        increment our count
    */
    intervals.sort((a,b) => {
        if (a[0] > b[0]){
            return 1
        }
        else if (a[0] < b[0]){
            return -1
        }
        return 0
    })
    let lastNonOverlapping = intervals[0]
    let res = 0
    for (let i = 1; i < intervals.length; ++i){
        let [prevStart, prevEnd] = lastNonOverlapping
        let [curStart, curEnd] = intervals[i]
        if (prevEnd > curStart){
            ++res
            lastNonOverlapping = curEnd < prevEnd ? intervals[i] : lastNonOverlapping
        }
        // if the last interval did not overlap, then we set the current interval to be the 
        // new "non-overlapping" interval
        else {
            lastNonOverlapping = intervals[i]
        }
    }
    return res
};